BPA/Ch.11/ipm.stan

```diff
@@ -0,0 +1,181 @@
+    }
+    return negative_infinity();
+  }
+}
+
+data {
+  int nyears;                      // Number of years
+  vector[nyears] y;                // Population counts
+  int J[nyears-1];                 // Total number of nestings recorded
+  int R[nyears-1];                 // Annual number of surveyed broods
+  int m[2*(nyears-1), nyears];     // m-array
+}
+
+parameters {
+  real<lower=0> sigma_y;           // Observation error
+  vector<lower=0>[nyears] N1;      // Number of 1-year juveniles
+  vector<lower=0>[nyears] Nad;     // Number of adults
+  real<lower=0,upper=1> mean_sjuv; // Mean survival prob. juveniles
+  real<lower=0,upper=1> mean_sad;  // Mean survival prob. adults
+  real<lower=0,upper=1> mean_p;    // Mean recapture prob.
+  real<lower=0> mean_fec;          // Mean productivity
+}
+
+transformed parameters {
+  vector<lower=0,upper=1>[nyears-1] sjuv;
+  vector<lower=0,upper=1>[nyears-1] sad;
+  vector<lower=0,upper=1>[nyears-1] p;
+  vector<lower=0>[nyears-1] f;
+  vector<lower=0>[nyears] Ntot;
+  simplex[nyears] pr[2*(nyears-1)];
+  vector<lower=0>[nyears-1] rho;
+
+  // Survival and recapture probabilities, as well as productivity
+  for (t in 1:(nyears - 1)) {
+    sjuv[t] <- mean_sjuv;
+    sad[t] <- mean_sad;
+    p[t] <- mean_p;
+    f[t] <- mean_fec;
+  }
+
+  // Total number of individuals
+  for (t in 1:nyears)
+    Ntot[t] <- Nad[t] + N1[t];
+
+  { // m-array
+      vector[nyears-1] q;
+
+      q <- 1.0 - p;
+
+      // m-array cell probabilities for juveniles
+      for (t in 1:(nyears - 1)) {
+        // Main diagonal
+        pr[t, t] <- sjuv[t] * p[t];
+
+        // Above main diagonal
+        for (j in (t + 1):(nyears - 1))
+          pr[t, j] <- sjuv[t] * prod(sad[(t + 1):j])
+            * prod(q[t:(j - 1)]) * p[j];
+
+        // Below main diagonal
+        for (j in 1:(t - 1))
+          pr[t, j] <- 0.0;
+
+        // Last column: probability of non-recapture
+        pr[t,nyears] <- 1.0 - sum(pr[t, 1:(nyears - 1)]);
+      } //t
+
+      // m-array cell probabilities for adults
+      for (t in 1:(nyears - 1)) {
+
+        // Main diagonal
+        pr[t + nyears - 1, t] <- sad[t] * p[t];
+        // Above main diagonal
+        for (j in (t + 1):(nyears - 1))
+          pr[t + nyears - 1, j] <- prod(sad[t:j])
+            * prod(q[t:(j - 1)]) * p[j];
+
+        // Below main diagonal
+        for (j in 1:(t - 1))
+          pr[t + nyears - 1, j] <- 0.0;
+
+        // Last column
+        pr[t + nyears - 1, nyears] <- 1.0
+          - sum(pr[t + nyears - 1, 1:(nyears - 1)]);
+      } //t
+  }
+
+  // Productivity
+  for (t in 1:(nyears - 1))
+    rho[t] <- R[t] * f[t];
+}
+
+model {
+  // Priors
+  // Initial population sizes
+  // Constraints ensure truncated normal (> 0)
+  N1[1] ~ normal(100, 100);
+  Nad[1] ~ normal(100, 100);
+
+  mean_sjuv ~ uniform(0, 1);
+  mean_sad ~ uniform(0, 1);
+  mean_p ~ uniform(0, 1);
+  // Improper flat priors are implicitly used on sigma_y and mean_fec.
```

Comment:
Is it intentional that some uniform priors are explicitly written and others are left to be implied by constraints? I think it would be better to be consistent, but it's not a huge deal.